src/node_modules.cc

```diff
@@ -0,0 +1,412 @@
+#include "node_modules.h"
+#include <cstdio>
+#include "base_object-inl.h"
+#include "node_errors.h"
+#include "node_external_reference.h"
+#include "node_url.h"
+#include "util-inl.h"
+#include "v8-fast-api-calls.h"
+#include "v8-function-callback.h"
+#include "v8-primitive.h"
+#include "v8-value.h"
+#include "v8.h"
+
+#include "simdjson.h"
+
+namespace node {
+namespace modules {
+
+using v8::Array;
+using v8::Context;
+using v8::FunctionCallbackInfo;
+using v8::HandleScope;
+using v8::Isolate;
+using v8::Local;
+using v8::NewStringType;
+using v8::Object;
+using v8::ObjectTemplate;
+using v8::Primitive;
+using v8::String;
+using v8::Undefined;
+using v8::Value;
+
+#ifdef __POSIX__
+constexpr char kPathSeparator = '/';
+#else
+const char* const kPathSeparator = "\\/";
+#endif
+
+void BindingData::MemoryInfo(MemoryTracker* tracker) const {
+  // Do nothing
+}
+
+BindingData::BindingData(Realm* realm,
+                         v8::Local<v8::Object> object,
+                         InternalFieldInfo* info)
+    : SnapshotableObject(realm, object, type_int) {}
+
+bool BindingData::PrepareForSerialization(v8::Local<v8::Context> context,
+                                          v8::SnapshotCreator* creator) {
+  // Return true because we need to maintain the reference to the binding from
+  // JS land.
+  return true;
+}
+
+InternalFieldInfoBase* BindingData::Serialize(int index) {
+  DCHECK_IS_SNAPSHOT_SLOT(index);
+  InternalFieldInfo* info =
+      InternalFieldInfoBase::New<InternalFieldInfo>(type());
+  return info;
+}
+
+void BindingData::Deserialize(v8::Local<v8::Context> context,
+                              v8::Local<v8::Object> holder,
+                              int index,
+                              InternalFieldInfoBase* info) {
+  DCHECK_IS_SNAPSHOT_SLOT(index);
+  HandleScope scope(context->GetIsolate());
+  Realm* realm = Realm::GetCurrent(context);
+  BindingData* binding = realm->AddBindingData<BindingData>(holder);
+  CHECK_NOT_NULL(binding);
+}
+
+Local<Array> BindingData::PackageConfig::Serialize(Realm* realm) {
+  auto has_manifest = !realm->env()->options()->experimental_policy.empty();
+  auto isolate = realm->isolate();
+  const auto ToString = [isolate](std::string_view input) -> Local<Primitive> {
+    return String::NewFromUtf8(
+               isolate, input.data(), NewStringType::kNormal, input.size())
+        .ToLocalChecked();
+  };
+  Local<Value> values[7] = {
+      name.has_value() ? ToString(*name) : Undefined(isolate),
+      main.has_value() ? ToString(*main) : Undefined(isolate),
+      ToString(type),
+      imports.has_value() ? ToString(*imports) : Undefined(isolate),
+      exports.has_value() ? ToString(*exports) : Undefined(isolate),
+      has_manifest ? ToString(raw_json) : Undefined(isolate),
+      ToString(file_path),
+  };
+  return Array::New(isolate, values, 7);
+}
+
+std::optional<BindingData::PackageConfig> BindingData::GetPackageJSON(
```

Comment:
I would be tempted to just return a pointer to the value in cache (e.g., PackageConfig*). The std::optional does not buy you much here.

Reply:
I don’t know how big PackageConfig is, but it gets copied into std::optional even when it was in cache. I can propose alternative designs.